lib/mpl_toolkits/mplot3d/axes3d.py

```diff
@@ -1671,64 +1671,57 @@ def plot_surface(self, X, Y, Z, *args, **kwargs):
             fcolors = self._shade_colors_lightsource(Z, cmap, lightsource)
 
         polys = []
-        # Only need these vectors to shade if there is no cmap
-        if cmap is None and shade :
-            totpts = int(np.ceil(float(rows - 1) / rstride) *
-                         np.ceil(float(cols - 1) / cstride))
-            v1 = np.empty((totpts, 3))
-            v2 = np.empty((totpts, 3))
-            # This indexes the vertex points
-            which_pt = 0
+
+        # evenly spaced, and including both endpoints
+        row_inds = list(xrange(0, rows-1, rstride)) + [rows-1]
+        col_inds = list(xrange(0, cols-1, cstride)) + [cols-1]
 
 
-        #colset contains the data for coloring: either average z or the facecolor
+        #colset contains the sampled facecolors
         colset = []
-        for rs in xrange(0, rows-1, rstride):
-            for cs in xrange(0, cols-1, cstride):
+        for rs, rs_next in zip(row_inds[:-1], row_inds[1:]):
+            for cs, cs_next in zip(col_inds[:-1], col_inds[1:]):
                 ps = []
                 for a in (X, Y, Z):
-                    ztop = a[rs,cs:min(cols, cs+cstride+1)]
-                    zleft = a[rs+1:min(rows, rs+rstride+1),
-                              min(cols-1, cs+cstride)]
-                    zbase = a[min(rows-1, rs+rstride), cs:min(cols, cs+cstride+1):][::-1]
-                    zright = a[rs:min(rows-1, rs+rstride):, cs][::-1]
+                    # the edges of the projected quadrilateral
+                    # note we use pythons half-open ranges to avoid repeating
+                    # the corners
+                    ztop   = a[rs,             cs:cs_next   ]
+                    zleft  = a[rs:rs_next,     cs_next      ]
+                    zbase  = a[rs_next,        cs_next:cs:-1]
+                    zright = a[rs_next:rs:-1,  cs           ]
                     z = np.concatenate((ztop, zleft, zbase, zright))
                     ps.append(z)
 
-                # The construction leaves the array with duplicate points, which
-                # are removed here.
-                ps = list(zip(*ps))
-                lastp = np.array([])
-                ps2 = [ps[0]] + [ps[i] for i in xrange(1, len(ps)) if ps[i] != ps[i-1]]
-                avgzsum = sum(p[2] for p in ps2)
-                polys.append(ps2)
+                # ps = np.stack(ps, axis=-1)
+                ps = np.array(ps).T
```

Comment:
I wonder if there's some way of pre-allocating this with the lengths of `row_inds` and `col_inds`.

Reply:
No, because `ps` varies in length throughout the iteration

Reply:
Err, I might have had the wrong loop; the lengths of bits out of `X`, `Y`, Z` are not known?

Reply:
No. They correspond the perimeters of the "square" "cells", which are not always the same size - the cells look like:

```
+---+---+-+
|   |   | |
|   |   | |
+---+---+-+
|   |   | |
|   |   | |
+---+---+-+
|   |   | |
+---+---+-+
```

where the trailing edges contain non-uniformly sized cells

Reply:
I think we're talking about different things; I mean `ps` itself which should be uniform since you're calling `np.array` on it, no?

Reply:
Ah, I see. Yes, we could preallocate to `(2*(rs_next - rs) + 2*(cs_next - cs), 3)`

Reply:
Ok, here's the answer - preallocating doesn't help, because `concatenate` has no `out` argument until numpy 1.14